yo1k/tic_tac_toe/core/game.py

```diff
@@ -0,0 +1,282 @@
+        self.game_rounds = game_rounds
+        self.round = round_
+        assert len(players) == State.players_count(), f"{players}"
+        self.players = players
+        self.step = step
+        self.phase = phase
+        self.board = board
+
+    def turn(self) -> int:
+        """Returns `turn` - the index of the current `Player` in `State.players`.
+
+        `turn` is calculated in such a way as to alternate players order in different
+        `round` while keeping the same `Player` index in the `State.players`.
+        """
+        return (self.step + self.round) % 2
+
+    @staticmethod
+    def players_count() -> int:
+        return 2
+
+    def __str__(self):
+        return (
+                f"{type(self)}("
+                f"game_rounds={self.game_rounds},"
+                f"round={self.round},"
+                f"players={self.players},"
+                f"step={self.step},"
+                f"phase={self.phase},"
+                f"board={self.board})")
+
+
+class Action:
+    def __init__(self, surrender: bool, occupy: Optional[Cell], next_round: bool):
+        assert \
+            (surrender is True and occupy is None and next_round is False) \
+            or (surrender is False and occupy is not None and next_round is False) \
+            or (surrender is False and occupy is None and next_round is True), \
+            f"surrender={surrender}, occupy={occupy}, next_round={next_round}"
+        self.__surrender = surrender
+        self.__occupy = occupy
+        self.__start = next_round
+
+    @staticmethod
+    def new_surrender() -> Action:
+        return Action(True, None, False)
+
+    @staticmethod
+    def new_occupy(cell: Cell) -> Action:
+        return Action(False, cell, False)
+
+    @staticmethod
+    def new_start() -> Action:
+        return Action(False, None, True)
+
+    @property
+    def surrender(self):
+        return self.__surrender
+
+    @property
+    def occupy(self):
+        return self.__occupy
+
+    @property
+    def start(self):
+        return self.__start
+
+    def __str__(self):
+        if self.__surrender is True:
+            action = "surrender"
+        elif self.__occupy is not None:
+            action = f"occupy={self.__occupy}"
+        elif self.__start is True:
+            action = "start"
+        else:
+            assert False
+        return (
+                f"{type(self)}("
+                f"{action})")
+
+
+class ActionQueue(ABC):
+    @abstractmethod
+    def next(self) -> Optional[Action]:
+        pass
+
+
+class Logic:
+    """Game logic."""
+
+    def __init__(self, action_queues: Sequence[ActionQueue]):
+        """Indexes in `action_queues` correspond to indexes in `State.players`."""
+        assert len(action_queues) == State.players_count(), f"{action_queues}"
+        self.__action_queues = action_queues
+
+    def advance(self, state: State):
+        turn = state.turn()
+        action = self.__action_queues[turn].next()
```

Comment:
`action = self.__action_queues[state.turn()].next()` is more concise and is still perfectly readable.

Reply:
Done.